### DISCUSSION TITLE
avif file compression size huge differences

### DISCUSSION THREAD

**Discussion by suntong** [libvips/libvips#3773] [Q&A] [answered]
Trying to build libvips from source [myself](https://github.com/suntong/bimg/commit/b332700185b8f6160eb6eb1cc127af88e55ce66b), and running `vips copy smart-crop.jpg smart-crop.avif` using my own build executable, I'm getting:

```sh
$ ls -l smart-crop.*
-rw-r--r-- 1 root root 65955 Dec  2 03:18 smart-crop.avif
-rw-r--r-- 1 root root 79289 Dec  2 03:16 smart-crop.jpg
```

However, if running `vips copy smart-crop.jpg smart-crop.avif` using debian stable package executable, I'm getting:

```sh
$ ls -l smart-crop.*
-rw-r--r-- 1 root root 79289 Dec  1 23:02 smart-crop.jpg
-rw-r--r-- 1 root root 18533 Dec  1 23:02 smart-crop.avif
```

Why such huge differences in avif file compression size please?

PS. The `smart-crop.jpg` file is from https://raw.githubusercontent.com/h2non/imaginary/master/testdata/smart-crop.jpg

**Comment by jcupitt** [comment, marked as answer]
Hello again. I tried here and I see:

```
$ vips copy smart-crop.jpg x.avif
$ ls -l x.avif smart-crop.jpg 
-rw-rw-r-- 1 john john 79289 Nov 30 10:06 smart-crop.jpg
-rw-r--r-- 1 john john 18530 Dec  2 10:48 x.avif
```

That's with my own libvips, so I agree your build looks odd.

I had a quick look at your dockerfile:

- you are using a *very* old version of libheif, I would swap to something newer ... I'm on 1.16.2, but current is 1.17.5. That might be the cause of the difference
- you are building against imagemagick -- are you certain you need it? There are security implications if you are exposing this build to untrusted data
- you don't need `gobject-introspection`, `gtk-doc-tools`, `swig`, `libgif`, `libgsf-1`
- you probably don't need `libxml2-dev` (libvips doesn't use it, we're `expat`)
- I would use `libspng` instead of `libpng` -- it's faster and more secure
- don't use `matio`, it's not safe against untrusted data
- are you are you need `cfitsio`? I would remove it unless you are certain
- `openslide` is also probably unnecessary, unless you have a very strong need
- `pdfium` would probably be faster than `poppler`, and would not expose you to licensing issues
- libvips uses Highway now in preference to Orc, I would consider swapping to that

**Reply by jcupitt** [reply]
I updated my heroku dockerfile for 8.15, it might be useful:

https://github.com/jcupitt/docker-builds/blob/master/libvips-heroku22/Dockerfile

**Reply by suntong** [reply]
Thanks for the most comprehensive reply. Most of the dependencies like libpng or imagemagick are what `bimg` used to require, the only extra packages that I added, beside it _and your readme_ is, `cmake` and `gobject-introspection`, which were added because of build failures. For e.g., I tried to remove `gobject-introspection` from my docker and am getting back the following errors again:

```
...
Configuring config.h using configuration
Program g-ir-scanner found: YES (/usr/bin/g-ir-scanner)
Program glib-mkenums found: YES (/usr/bin/glib-mkenums)
Configuring version.h using configuration
Found pkg-config: /usr/bin/pkg-config (0.29.2)
Program glib-genmarshal found: YES (/usr/bin/glib-genmarshal)
Found CMake: /usr/bin/cmake (3.18.4)
Build-time dependency gobject-introspection-1.0 found: NO (tried pkgconfig and cmake)

libvips/meson.build:45:4: ERROR: Dependency "gobject-introspection-1.0" not found, tried pkgconfig and cmake
```

Maybe my base image is too old (or maybe `heroku:22` _already have it_).
Just that the `docker` I'm using cannot use recent containers:

```
$ docker run -it debian
root@3cfd4b1d726b:/# apt-get update
Get:1 http://deb.debian.org/debian bookworm InRelease [151 kB]
Get:2 http://deb.debian.org/debian bookworm-updates InRelease [52.1 kB]
Get:3 http://deb.debian.org/debian-security bookworm-security InRelease [48.0 kB]
Get:4 http://deb.debian.org/debian bookworm/main amd64 Packages [8780 kB]
Get:5 http://deb.debian.org/debian bookworm-updates/main amd64 Packages [6668 B]
Get:6 http://deb.debian.org/debian-security bookworm-security/main amd64 Packages [106 kB]
Fetched 9144 kB in 1s (7776 kB/s)                         
Reading package lists... Done
E: Problem executing scripts APT::Update::Post-Invoke 'rm -f /var/cache/apt/archives/*.deb /var/cache/apt/archives/partial/*.deb /var/cache/apt/*.bin || true'
E: Sub-process returned an error code
```

That `E: Problem executing scripts APT::Update::Post-Invoke`, based on my research, is that my docker is too old, yet I don't have the sudo right to update it. 

I'll look into your heroku dockerfile trying to update mine...

**Reply by jcupitt** [reply]
Pass `-Dintrospection=disabled` to libvuips meson to stop it looking for the introspection stuff.

**Reply by suntong** [reply]
Aha! Gotcha. Thanks.

**Reply by jcupitt** [reply]
Your base image is also extremely old, I think I would use `golang:1.21.4`.

**Reply by suntong** [reply]
Yeah, agree, just that I cannot do any recent versions with my current VPC because of the `E: Problem executing scripts APT::Update::Post-Invoke`. I'll try to find another build environment.

**Reply by suntong** [reply]
Hi again @jcupitt, 

> I tried here and I see:
> 
> ```
> $ vips copy smart-crop.jpg x.avif
> ```

what would the `quality` value used by ` vips` be _by default_ when saving `avif` files, like above?

I'm already using the latest libvips in [my imaginary](https://github.com/suntong/imaginary/tree/dev):

```
$ curl http://localhost:9000/
{"imaginary":"dev","bimg":"1.1.9","libvips":"8.15.0"}
```

But when converting from `smart-crop.jpg` above, the converted size is 47083 (from the original 79289), 

`curl -O "http://localhost:9000/convert?type=avif&url=https://raw.githubusercontent.com/h2non/imaginary/master/testdata/smart-crop.jpg"`

when I tried with `quality=60`:

`curl -O "http://localhost:9000/convert?type=avif&quality=60&url=https://raw.githubusercontent.com/h2non/imaginary/master/testdata/smart-crop.jpg"`

The saved size is 29977, still not close to the desired 18530 size. 

Hence I want to know what the `quality` value to use to overwrite the default in
https://github.com/suntong/bimg/blob/master/resizer.go#L193-L209

Thanks

**Reply by jcupitt** [reply]
AVIF doesn't have an official default quality setting. I think we've changed the one libvips uses a few times, but it's currently 50.

```
$ vips heifsave
save image in HEIF format
usage:
   heifsave in filename [--option-name option-value ...]
where:
   in           - Image to save, input VipsImage
   filename     - Filename to save to, input gchararray
optional arguments:
   Q            - Q factor, input gint
			default: 50
			min: 1, max: 100
   bitdepth     - Number of bits per pixel, input gint
			default: 12
			min: 1, max: 16
   lossless     - Enable lossless compression, input gboolean
			default: false
...
```

**Reply by suntong** [reply]
Oh, thanks for the reply, with `quality=50`, the imaginary produces the size of 18441, quite close to the desired 18530 size. Perfect!

Case closed! 

Thanks
